apps/manager_app/device_authentication.c

```diff
@@ -0,0 +1,261 @@
+    /* Fetch and encode serial signature, send it to the host and wait for
+     * further USB requests. */
+    manager_result_t result = MANAGER_RESULT_INIT_ZERO;
+    result.which_response = MANAGER_RESULT_AUTH_DEVICE_TAG;
+    result.auth_device.which_response =
+        MANAGER_AUTH_DEVICE_SERIAL_SIG_RESPONSE_SERIAL_TAG;
+    result.auth_device = sign_serial_number();
+
+    uint8_t response[DEVICE_AUTH_RESPONSE_SIZE] = {0};
+    size_t bytes_encoded = 0;
+    encode_manager_result(
+        &result, &response[0], sizeof(response), &bytes_encoded);
+    usb_send_msg(&response[0], bytes_encoded);
+
+    /* Set the next state to SIGN_RANDOM_NUM */
+    *state_ptr = SIGN_RANDOM_NUM;
+  }
+
+  return;
+}
+
+static void device_auth_usb_cb(usb_event_t usb_evt,
+                               device_auth_flow_states_e *state_ptr) {
+  device_auth_flow_states_e state = *state_ptr;
+
+  /* Decode recieved query using protobuf helpers */
+  manager_query_t query = MANAGER_QUERY_INIT_ZERO;
+  if (false == decode_manager_query(usb_evt.p_msg, usb_evt.msg_size, &query)) {
+    // TODO: Handle proto decode error code
+    // TODO: Early flow exit
+    return;
+  }
+
+  /* Decode recieved request_type */
+  pb_size_t request_type = get_request_type(
+      (const manager_auth_device_request_t *)&(query.auth_device));
+
+  if (SIGN_RANDOM_NUM == state) {
+    if (MANAGER_AUTH_DEVICE_REQUEST_CHALLENGE_TAG == request_type) {
+      /* Fetch and encode random challenge signature, send it to the host and
+       * wait for further USB requests. */
+      manager_result_t result = MANAGER_RESULT_INIT_ZERO;
+      result.which_response = MANAGER_RESULT_AUTH_DEVICE_TAG;
+      result.auth_device.which_response =
+          MANAGER_AUTH_DEVICE_SERIAL_SIG_RESPONSE_SIGNATURE_TAG;
+      result.auth_device =
+          sign_random_challenge(&(query.auth_device.challenge.challenge[0]));
+
+      uint8_t response[DEVICE_AUTH_RESPONSE_SIZE] = {0};
+      size_t bytes_encoded = 0;
+      encode_manager_result(
+          &result, &response[0], sizeof(response), &bytes_encoded);
+      usb_send_msg(&response[0], bytes_encoded);
+      *state_ptr = RESULT;
+    } else if (MANAGER_AUTH_DEVICE_REQUEST_RESULT_TAG == request_type) {
+      /* If SIGN_RANDOM_NUM == state, but the request_type received is
+       * MANAGER_AUTH_DEVICE_REQUEST_RESULT_TAG, it's an unexpected step in
+       * the flow. The device will treat it as an attempt to force device
+       * authentication status */
+      device_auth_handle_response(false);
+      *state_ptr = FLOW_COMPLETE;
+      usb_clear_event();
+      // TODO: Early flow exit
+    } else {
+      // TODO: Handle error scenario in which an unexpected request_type was
+      // detected
+      // TODO: Early flow exit
+    }
+  } else if (RESULT == state) {
+    /* Handle verification status */
+    bool verified = query.auth_device.result.verified;
+    device_auth_handle_response(verified);
+    usb_clear_event();
```

Comment:
the `device_auth_handle_response` renders delay screens. delay screens pause the execution. if `usb_clear_event` is called after that, the host gets an invalid state if it does not entertain `status` responses. When testing with bare-bone SDK, new commands fail with error as the buffer is blocked till `usb_clear_event`

Reply:
look at other places too eg: https://github.com/Cypherock/x1_wallet_firmware/pull/254/files#diff-2b0c4f85f60d96db37e84ce447207de7ff219c9087b6cd5578eed2d7f43c180aR203-R205

Reply:
Hi Ujjwal, as per our discussion with Irshad.. We have added a dummy message from device to host to signal flow completion.

Host will poll for the completion message before starting any other transaction.

Reply:
makes sense. That makes the communication consistent.